server/store/session.go

```diff
@@ -0,0 +1,170 @@
+	DeleteAllForApp(appID apps.AppID) error
+}
+
+type sessionStore struct {
+	*Service
+}
+
+var _ SessionStore = (*sessionStore)(nil)
+
+func sessionKey(appID apps.AppID, userID string) string {
+	return appKey(appID) + "_" + userID
+}
+
+func appKey(appID apps.AppID) string {
+	return config.KVTokenPrefix + "_" + string(appID)
+}
+
+func parseKey(key string) (apps.AppID, string, error) {
+	s := strings.Split(key, "_")
+	if len(s) != 3 {
+		return "", "", errors.New("invalid key pattern")
+	}
+
+	if s[0] != config.KVTokenPrefix {
+		return "", "", errors.New("invalid key prefix")
+	}
+
+	return apps.AppID(s[1]), s[2], nil
+}
+
+func (s sessionStore) Get(appID apps.AppID, userID string) (*model.Session, error) {
+	return s.get(sessionKey(appID, userID))
+}
+
+func (s sessionStore) get(key string) (*model.Session, error) {
+	var session model.Session
+	err := s.conf.MattermostAPI().KV.Get(key, &session)
+	if err != nil {
+		return nil, err
+	}
+
+	if session.Id == "" {
+		return nil, utils.ErrNotFound
+	}
+
+	return &session, nil
+}
+
+func (s sessionStore) Save(appID apps.AppID, userID string, session *model.Session) error {
+	_, err := s.conf.MattermostAPI().KV.Set(sessionKey(appID, userID), session)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s sessionStore) listKeysForApp(appID apps.AppID) ([]string, error) {
+	ret := make([]string, 0)
+
+	for i := 0; ; i++ {
+		keys, err := s.conf.MattermostAPI().KV.ListKeys(i, keysPerPage, pluginapi.WithPrefix(appKey(appID)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list keys - page, %d", i)
+		}
+
+		ret = append(ret, keys...)
+
+		if len(keys) < keysPerPage {
+			break
+		}
+	}
+
+	return ret, nil
+}
+
+func (s sessionStore) ListForApp(appID apps.AppID) ([]*model.Session, error) {
+	keys, err := s.listKeysForApp(appID)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]*model.Session, 0)
+
+	for _, k := range keys {
+		session, err := s.get(k)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed get session for key, %s", k)
+		}
+
+		ret = append(ret, session)
+	}
+
+	return ret, nil
+}
+
+func (s sessionStore) ListForUser(userID string) ([]*model.Session, error) {
+	ret := make([]*model.Session, 0)
+
+	for i := 0; ; i++ {
+		keys, err := s.conf.MattermostAPI().KV.ListKeys(i, keysPerPage)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list keys - page, %d", i)
+		}
+
+		for _, k := range keys {
+			_, keyUserID, err := parseKey(k)
+			if err != nil {
+				continue
+			}
+
+			if keyUserID != userID {
+				continue
+			}
+
+			session, err := s.get(k)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed get session for key, %s", k)
+			}
+
+			ret = append(ret, session)
+		}
+
+		if len(keys) < keysPerPage {
+			break
+		}
+	}
+
+	return ret, nil
+}
+
+func (s sessionStore) DeleteAllForApp(appID apps.AppID) error {
+	keys, err := s.listKeysForApp(appID)
+	if err != nil {
+		return err
+	}
+
+	for _, k := range keys {
```

Comment:
nist: s/k/key (k is typically used to denote diff types of vars - like `i`, 'j`, etc.)